huaweicloud/services/ces/resource_huaweicloud_ces_dashboard_widget.go

```diff
@@ -0,0 +1,541 @@
+				Computed:    true,
+				Description: `Specifies how metric data is aggregated.`,
+			},
+			"order": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: `Specifies how top n resources by a metric are sorted on a dashboard widget.`,
+			},
+		},
+	}
+	return &sc
+}
+
+func widgetDimensionOptsSchema() *schema.Resource {
+	sc := schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Specifies the dimension name.`,
+			},
+			"filter_type": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Specifies the resource type.`,
+			},
+			"values": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: `Specifies the dimension value list.`,
+			},
+		},
+	}
+	return &sc
+}
+
+func resourceDashboardWidgetCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	var (
+		createWidgetHttpUrl = "v2/{project_id}/dashboards/{dashboard_id}/widgets"
+		createWidgetProduct = "ces"
+	)
+	createWidgetClient, err := cfg.NewServiceClient(createWidgetProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating CES client: %s", err)
+	}
+
+	createWidgetPath := createWidgetClient.Endpoint + createWidgetHttpUrl
+	createWidgetPath = strings.ReplaceAll(createWidgetPath, "{project_id}", createWidgetClient.ProjectID)
+	createWidgetPath = strings.ReplaceAll(createWidgetPath, "{dashboard_id}", d.Get("dashboard_id").(string))
+
+	createWidgetOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		MoreHeaders:      map[string]string{"Content-Type": "application/json;charset=UTF-8"},
+	}
+	params := utils.RemoveNil(buildCreateWidgetBodyParams(d))
+	createWidgetOpt.JSONBody = []map[string]interface{}{params}
+	createWidgetResp, err := createWidgetClient.Request("POST", createWidgetPath, &createWidgetOpt)
+	if err != nil {
+		return diag.Errorf("error creating CES dashboard widget: %s", err)
+	}
+
+	createWidgetRespBody, err := utils.FlattenResponse(createWidgetResp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := jmespath.Search("widget_ids|[0]", createWidgetRespBody)
+	if err != nil {
+		return diag.Errorf("error creating CES dashboard widget: ID is not found in API response")
+	}
+	d.SetId(id.(string))
+
+	if _, ok := d.GetOk("unit"); ok {
+		updateWidgetPath := createWidgetClient.Endpoint + "v2/{project_id}/widgets/batch-update"
+		updateWidgetPath = strings.ReplaceAll(updateWidgetPath, "{project_id}", createWidgetClient.ProjectID)
+		updateWidgetOpt := golangsdk.RequestOpts{
+			KeepResponseBody: true,
+			MoreHeaders:      map[string]string{"Content-Type": "application/json;charset=UTF-8"},
+		}
+		param := map[string]interface{}{
+			"widget_id": d.Id(),
+			"unit":      d.Get("unit"),
+		}
```

Comment:
how about this, no need to get again.
```suggestion
	if v, ok := d.GetOk("unit"); ok {
		updateWidgetPath := createWidgetClient.Endpoint + "v2/{project_id}/widgets/batch-update"
		updateWidgetPath = strings.ReplaceAll(updateWidgetPath, "{project_id}", createWidgetClient.ProjectID)
		updateWidgetOpt := golangsdk.RequestOpts{
			KeepResponseBody: true,
			MoreHeaders:      map[string]string{"Content-Type": "application/json;charset=UTF-8"},
		}
		param := map[string]interface{}{
			"widget_id": d.Id(),
			"unit":      v,
		}
```